L1Trigger/L1TCaloLayer1/src/UCTGeometry.cc

```diff
@@ -0,0 +1,229 @@
+  }
+  return uctRegionPhiIndex;
+}
+
+uint32_t UCTGeometry::getCrate(int caloEta, int caloPhi) {
+  uint32_t crate = 0xDEADBEEF;
+  if(caloPhi >= 11 && caloPhi <= 34) crate = 0;
+  else if(caloPhi >= 35 && caloPhi <= 58) crate = 2;
+  else if(caloPhi >= 59 && caloPhi <= 72) crate = 1;
+  else if(caloPhi >= 1 && caloPhi <= 10) crate = 1;  
+  return crate;
+}
+
+uint32_t UCTGeometry::getCard(int caloEta, int caloPhi) {
+  uint32_t crate = getCrate(caloEta, caloPhi);
+  uint32_t card = 0xDEADBEEF;
+  if(crate == 0) {
+    card = (caloPhi - 11) / 4;
+  }
+  else if(crate == 2) {
+    card = (caloPhi - 35) / 4;
+  }
+  else if(crate == 1 && caloPhi > 58) {
+    card = (caloPhi - 59) / 4;
+  }
+  else if(crate == 1 && caloPhi <= 10) {
+    card = (caloPhi + 13) / 4;
+  }    
+  return card;
+}
+
+uint32_t UCTGeometry::getRegion(int caloEta, int caloPhi) {
+  uint32_t absCEta = abs(caloEta);
+  if((absCEta - 1) < (NRegionsInCard * NEtaInRegion))
+    return (absCEta - 1) / NEtaInRegion;
+  else
+    return NRegionsInCard + ((absCEta - 2 - (NRegionsInCard * NEtaInRegion)) / NHFEtaInRegion);
+}
+
+uint32_t UCTGeometry::getiEta(int caloEta) {
+  uint32_t absCEta = abs(caloEta);
+  if((absCEta - 1) < (NRegionsInCard * NEtaInRegion))
+    return (absCEta - 1) % NEtaInRegion;
+  else
+    return absCEta % NHFEtaInRegion;  // To account for missing tower 29
+}
+
+uint32_t UCTGeometry::getiPhi(int caloPhi) {
+  return (caloPhi + 1) % NPhiInCard;
+}
+
+uint32_t UCTGeometry::getNEta(uint32_t region) {
+  uint32_t nEta = 0xDEADBEEF;
+  if(region < CaloHFRegionStart) {
+    nEta = NEtaInRegion;
+  }
+  else {
+    nEta = NHFEtaInRegion;
+  }
+  return nEta;
+}
+
+uint32_t UCTGeometry::getNPhi(uint32_t region) {
+  return NPhiInRegion;
+}
+
+UCTRegionIndex UCTGeometry::getUCTRegionIndex(int caloEta, int caloPhi) {
+  uint32_t regionPhi = getUCTRegionPhiIndex(getCrate(caloEta, caloPhi), getCard(caloEta, caloPhi));
+  int regionEta = getUCTRegionEtaIndex((caloEta < 0), getRegion(caloEta, caloPhi));
+  return UCTRegionIndex(regionEta, regionPhi);
+}
+
+UCTRegionIndex UCTGeometry::getUCTRegionIndex(bool negativeSide, uint32_t crate, uint32_t card, uint32_t region) {
+  uint32_t regionPhi = getUCTRegionPhiIndex(crate, card);
+  int regionEta = getUCTRegionEtaIndex(negativeSide, region);
+  return UCTRegionIndex(regionEta, regionPhi);
+}
+
+UCTTowerIndex UCTGeometry::getUCTTowerIndex(UCTRegionIndex region, uint32_t iEta, uint32_t iPhi) {
+  if(iPhi >= NPhiInRegion || iEta >= NEtaInRegion) {
+    return UCTTowerIndex(0, 0); // Illegal values
+  }
+  int regionEta = region.first;
+  int absRegionEta = abs(regionEta);
+  int towerEta = (regionEta / absRegionEta) * (absRegionEta * getNEta(absRegionEta) + iEta);
+  uint32_t regionPhi = region.second;
+  int towerPhi = regionPhi * getNPhi(absRegionEta) + iPhi;
+  return UCTTowerIndex(towerEta, towerPhi);
+}
+
+double UCTGeometry::getUCTTowerEta(int caloEta) {
+  static bool first = true;
+  static double twrEtaValues[29];
```

Comment:
put this array into the member data (not as static) and its initialization into the constructor